### PR TITLE
P3142R0: Printing Blank Lines with `println`

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements:  https://wg21.link/P3142R0 (Fixes GH-4524)

Applied retroactively as DR:
https://github.com/orgs/microsoft/projects/1143?pane=issue&itemId=57457187

### CODE REVIEW

**Comment by H-G-Hristov** [comment]
@microsoft-github-policy-service agree

**Comment by StephanTLavavej** [comment]
Thanks! I see that this is marked as a draft PR - is it ready for review? When PRs are marked as drafts, the PR system doesn't run checks for them.

Also, you should edit your PR description to say "Fixes GH-nnnn" or "Fixes #nnnn" to properly link the PR to the issue in a way that merging the PR will auto-close the issue (only the PR description can do this; the PR title, or following comments, don't count; I could edit it myself but I thought I'd mention how it works).

**Comment by H-G-Hristov** [comment]
> Thanks! I see that this is marked as a draft PR - is it ready for review? When PRs are marked as drafts, the PR system doesn't run checks for them.
> 
> Also, you should edit your PR description to say "Fixes GH-nnnn" or "Fixes #nnnn" to properly link the PR to the issue in a way that merging the PR will auto-close the issue (only the PR description can do this; the PR title, or following comments, don't count; I could edit it myself but I thought I'd mention how it works). See the docs for [GitHub's close/fix/resolve syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

Thanks for the instruction! I think the PR is ready for a review if the tests pass.

I am unable to run the tests locally. I believe I followed the instructions correctly yet I get the following error (BTW X: is a subst-ed drive):

```
X:\GitHub\Microsoft\STL\out\x64>python tests\utils\stl-lit\stl-lit.py X:\GitHub\Microsoft\STL\tests\std\tests\P2093R14_formatted_output\test.cpp
stl-lit.py: X:\GitHub\Microsoft\STL\tests\std\lit.cfg:4: fatal: You seem to be running Lit directly -- you should be running Lit through <build>/tests/utils/stl-lit/stl-lit.py, which will ensure that the right Lit configuration file is used.
```

So I hoped I could run them through the CI.

**Comment by StephanTLavavej** [comment]
Thanks - you seem to be in the right directory running the right file, so perhaps it's the exotic subst configuration that's causing problems - I'll bring it up on our Discord's testing channel to see if anyone has any ideas.

I've switched your PR over to "ready to review" and manually queued a CI run (note that only maintainers can use that Azure Pipelines comment-command). Pushing further commits will automatically queue CI runs, it's just that the transition from draft to non-draft doesn't itself queue anything.

Our CI system is shared across all contributors and maintainers. It consumes internal budget that normally we don't care *at all* about, but if it crosses a certain threshold then internal people will start hissing at us (me), so we generally ask contributors to be considerate about setting clang-format-on-save in their editors and running a subset of tests locally before pushing, as opposed to rapid-fire pushes that stress the CI for validation. That said, since local tests are messed up for you, don't worry about relying on the CI here.

I should have time to review this week!

**Comment by StephanTLavavej** [comment]
Since we've confirmed that the `subst` is the cause, until we find a way to fix it, you should be able to run the tests locally if you use the true location of your repo. Thanks for bringing this to our attention!

**Comment by StephanTLavavej** [comment]
Comparing this test run to a successful test run, it appears that multiple configurations of `P2093R14_formatted_output` hung, which is why every x64 shard failed (and so I won't attempt a rerun).

File: tests/std/include/test_header_units_and_modules.hpp
Comment (StephanTLavavej) [inline tests/std/include/test_header_units_and_modules.hpp:549]: I figured out what was causing the test coverage to hang in the CI! `P2093R14_formatted_output` does very special things to create a console without popping up windows during a test run, and it interacts with lit in such a way that the test hangs if any assertion failures occur. These asserts are visible if the test is directly compiled and run. (The problems were things like expecting newlines to be embedded in strings, when the `get_console_line` machinery will perceive that line number ending.)

I was able to restore test coverage - some exactly captured from your original commit, other parts reworked.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! :heart_eyes_cat: I pushed tiny changes to the product code, and then I figured out how to add test coverage without it hanging.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by Zingam** [comment]
@StephanTLavavej  I was quite busy and I wasn't able to respond earlier. 
I pushed some changes without checking your latest comments first.  Sorry about that. Should I revert those?

Basically the idea is to test `println()` alone. In libc++ I did that for other another reason though.

Thank you!

**Comment by H-G-Hristov** [comment]
@StephanTLavavej  I was quite busy and I wasn't able to respond earlier. 
I pushed some changes without checking your latest comments first.  Sorry about that. Should I revert those?

Basically the idea is to test `println()` alone. In libc++ I did that for other another reason though.

Thank you!

**Comment by Zingam** [comment]
> @StephanTLavavej I was quite busy and I wasn't able to respond earlier. I pushed some changes without checking your latest comments first. Sorry about that. Should I revert those?
> 
> Basically the idea is to test `println()` alone. In libc++ I did that for other another reason though.
> 
> Thank you!

I reverted my changes.

**Comment by H-G-Hristov** [comment]
> @StephanTLavavej I was quite busy and I wasn't able to respond earlier. I pushed some changes without checking your latest comments first. Sorry about that. Should I revert those?
> 
> Basically the idea is to test `println()` alone. In libc++ I did that for other another reason though.
> 
> Thank you!

I reverted those changes.

**Comment by StephanTLavavej** [comment]
No worries, and thanks for the revert! I've already mirrored this to the MSVC-internal repo (stacked up with 20 other commits), so further changes would have reset that process.

If you'd like to add additional test coverage, followup PRs are always welcome after this has merged to `main`.

Of course, if you discover late-breaking correctness issues, it's always possible to rework or hold back a PR that's already been mirrored, we just like to avoid disrupting the process for "nice to have" changes.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks again for implementing this feature, which is sure to be used by programmers from beginners to experts everywhere, and congratulations on your first microsoft/STL commit! :heart_eyes_cat: :tada: :rocket:

This is expected to ship in VS 2022 17.11 Preview 2.

**Comment by H-G-Hristov** [comment]
> Thanks again for implementing this feature, which is sure to be used by programmers from beginners to experts everywhere, and congratulations on your first microsoft/STL commit! 😻 🎉 🚀
> 
> This is expected to ship in VS 2022 17.11 Preview 2.

Thank you very much for your support!

**Pull request merged**
